File: core/FirstMissingPositive/FirstMissingPositive.java
package com.dbc;

public class FirstMissingPositive {
    public int firstMissingPositive(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] <= 0) nums[i] = nums.length + 3;
        }
        for (int i = 0; i < nums.length; i++) {
            int num = Math.abs(nums[i]);
            if (num <= nums.length) nums[num - 1] = -Math.abs(nums[num - 1]);
        }
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > 0) return i + 1;
        }
        return nums.length + 1;
    }
}

File: core/FirstMissingPositive/FirstMissingPositive.py
from typing import List

def firstMissingPositive(nums: List[int]) -> int:
    length = len(nums)
    for i in range(length):
        if nums[i] <= 0:
            nums[i] = length + 1

    for i in range(length):
        num = abs(nums[i])
        if num <= length:
            nums[num - 1] = -abs(nums[num - 1])

    for i in range(length):
        if nums[i] > 0:
            return i + 1

    return length + 1


if __name__ == '__main__':
    print(firstMissingPositive([3,4,-1,1]))

